### ISSUE TITLE
PDE editor's code does not update when sketch is modified in an external editor

### ISSUE THREAD

**Issue by vivekmalhotra** [processing/processing#5466]
## Description

When a sketch is opened in the Processing Development Environment (PDE) and is then opened/modified in an external editor the change is not reflected in the PDE editor.   This occurs on my Windows 10, 64-bit machine only. On my other laptop, Windows 7 64-bit, upon changing in an external editor (e.g. VS code or Sublime Text 3) and saving the changes there pops a message in PDE indicating the file has changed - do you want to reload?

In Windows 10 64-bit this does not occur.  When the file is modified & saved in the external editor and we go back to the PDE, the file does not show/display updated code.  BUT, the PDE editor definitely is verifying the code and showing any errors but randomly.

The issue is that although the original file has been modified and the PDE editor.watch setting is set to be true, it just does not display the changed code.

## Steps to Reproduce

1. Open a sketch file in PDE editor in Windows 10 64-bit machine
2. Open the same file in another editor and make some changes to it in this external editor and save it 
3. The change is detected by the PDE editor environment but the code is not updated
4. Please refer to the 2 pictures attached below.  I opened the same file in PDE editor & then in VS code. Added a code line error_code() in VS code and saved the file.  In the second pic you can see that PDE editor log does show that it detected code modification and the code validation detects the error and in the suggestion shows that error_code() function does not exist - even though this code is not shown in the PDE editor. 

## Your Environment

 

* Processing version: v3.3.7
* Operating System and OS version: Windows 10 64-Bit
* Other information: 

## Possible Causes / Solutions

![vs_code_change](https://user-images.githubusercontent.com/4213286/38529780-3d3dcf2a-3c24-11e8-9513-ff2e90d98342.PNG)

![pde_code_unchanged](https://user-images.githubusercontent.com/4213286/38529784-3ed57496-3c24-11e8-9b8c-1d744b0fc5c9.PNG)

**Comment by kfrajer** [comment]
This needs some consideration. Saving the file on disk while is open in the PDE is sort of dangerous. The reason is that currently, you can load the files in the PDE and after that point, the files in the PDE is **not** guaranteed to be the same as the one  on disk. If you make a change using VS code for instance, you might have made some changes in parallel on the loaded sketch in the PDE, which are not stored in disk automatically. Trying to reload the files from disk (due to changes from VS code) could conflict with unsaved changes in the current PDE session. This requires careful consideration. Nevertheless this is a bug. 

For instance, right now I can open a sketch, modify it, not save changes and run it. The PDE runs the modified version without saving the changes. In fact, I can run for hours continuously modifying the sketch without saving it. If I close the PDE, the PDE will prompt to save the changes. In other words, the PDE is managing the current loaded sketch and it is not aware of changes of files on disk. 


If checking the file on disk is important or desired, then one approach would be to check for the file states every time the PDE gets focus. A second approach is to implement an auto-save feature which will check if the file(s) to be saved has not been modified on disk. This auto-save feature does not exist in Processing. Right now, the user has the onus to save their own work. 

Kf

**Comment by vivekmalhotra** [comment]
@kfrajer  Using PDE Editor on Windows 7 (64-bit) machine, it is aware of the changes in the file & prompts user to reload the file.  

Even in Windows 10 machine, if the file was to be modified outside of the PDE editor (e.g. VS code) it has detected that change and reloaded the file in the editor (if you click run in PDE editor, it will run the modified file).    Here is where the issue lies - even though the PDE editor has detected & read the file changes, it does not display the modified code in the editor (as I explained in my original post).

In summary, PDE Editor is aware of changes in the currently open file that was modified in an external editor.  It runs the modified version of the file.    The issue is that it does not display the modified version of the file in its own editor.  Or at least, should prompt the user whether it should display the modified file.    It is doing half the job and that is dangerous.

**Comment by JakubValtar** [comment]
Thanks for reporting this. There was a bug where the editor would not refresh the textarea if the sketch had only one tab. Should be fixed in the PR.
